WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/StatsPage.java

```diff
@@ -0,0 +1,179 @@
+                            )
+                    ),
+                    hasDescendant(allOf(
+                            withText("Comments"),
+                            hasSibling(withText(visitsData.getComments()))
+                            )
+                    )
+            ));
+
+            cardStructure.check(matches(isCompletelyDisplayed()));
+            return this;
+    }
+
+    public void assertKeyValuePairs(String cardHeader, List<StatsKeyValueData> list) {
+        for (StatsKeyValueData item : list) {
+            // Element with ID = stats_block_list
+            // |--Has child with text: e.g. "Posts and Pages"
+            // |--Has descendant that both:
+            //    |- Has text: post.title
+            //    |- Has a sibling with post.views (which means they're shown on same row):
+            ViewInteraction cardStructure = onView(allOf(
+                    withId(R.id.stats_block_list),
+                    hasDescendant(withText(cardHeader)),
+                    hasDescendant(allOf(
+                            withText(item.getKey()),
+                            hasSibling(withText(item.getValue()))
+                            )
+                    )
+            ));
+
+            cardStructure.check(matches(isCompletelyDisplayed()));
+        }
+    }
+
+    public StatsPage assertPosts(List<StatsKeyValueData> list) {
+        assertKeyValuePairs("Posts and Pages", list);
+        return this;
+    }
+
+    public StatsPage assertReferrers(List<StatsKeyValueData> list) {
+        assertKeyValuePairs("Referrers", list);
+        return this;
+    }
+
+    public StatsPage assertClicks(List<StatsKeyValueData> list) {
+        assertKeyValuePairs("Clicks", list);
+        return this;
+    }
+
+    public StatsPage assertAuthors(List<StatsKeyValueData> list) {
+        assertKeyValuePairs("Authors", list);
+        return this;
+    }
+
+    public StatsPage assertCountries(List<StatsKeyValueData> list) {
+        assertKeyValuePairs("Countries", list);
+        return this;
+    }
+
+    public StatsPage assertVideos(List<StatsKeyValueData> list) {
+        assertKeyValuePairs("Videos", list);
+        return this;
+    }
+
+    public StatsPage assertDownloads(List<StatsKeyValueData> list) {
+        assertKeyValuePairs("File downloads", list);
+        return this;
+    }
+
+    private void scrollToCard(String cardHeader) {
+        // A card that has "stats_block_list"
+        // and additionally contains a descendant
+        // with needed text:
+        ViewInteraction card = onView(allOf(
+                withId(R.id.stats_block_list),
+                hasDescendant(withText(cardHeader))
+                )
+        );
+
+        scrollIntoView(card);
+    }
+
+    public static void scrollIntoView(ViewInteraction container) {
+        int swipeCount = 0;
+        while (!isElementCompletelyDisplayed(container) && swipeCount < 20) {
+            swipeUpOnView(R.id.statsPager, (float) 0.5);
```

Comment:
I changed the `yFactor` from 1 to 0.5 here because on CI the single swipe distance would be so big that larger cards, e.g. `Countries`, would actually never be fully displayed, not because they don't fit the screen, but because the test swiped too far down, and either the card top or bottom would always be the outside visible area.

Reply:
Not suggesting to do it now, but `scrollIntoView(ViewInteraction)` is becoming a good candidate moving somewhere like `WPSupportUtils` as we have pretty much the same implementation in `ReaderPage`.

https://github.com/wordpress-mobile/WordPress-Android/blob/60ce7935b33fab0eee1c62aa174681105de53e6e/WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/ReaderPage.java#L37-L43

Reply:
Thanks for the suggestion! That was my initial plan, but then I got stopped by the fact that we will need to pass one more argument, which is an object that should be scrolled. Anyway, I just did that in bc4d0cf90329b628d6c288c421f4c28a15ba3f77, and actually also had to add a third parameter - the `yFactor`. Reader tests fail if it's 0.5, and Stats test fails if it's 1. So it's a parameter now as well. It's still less code in the end.